alica_engine/include/engine/blackboard/Blackboard.h

```diff
@@ -23,25 +54,143 @@ class BlackboardImpl
+        if (entry[Strings::key].as<std::string>() != key) {
+            // value is not predefined in pml file
+            vals.at(key) = value;
+        } else {
+            // value is predefined in pml file
+            std::string inputType = getNameFromType<T>();
+            std::string pmlType = getBlackboardValueType(key);
+            if (pmlType == inputType) {
+                vals.at(key) = value;
+            } else {
+                // types dont match, throw exception
+                throw BlackboardTypeMismatch(inputType, pmlType);
+            }
+        }
+    }
 
     bool hasValue(const std::string& key) const { return vals.count(key); }
     void removeValue(const std::string& key) { vals.erase(key); }
 
+    void initDefaultValues();
+
     void clear() { vals.clear(); }
     bool empty() const { return vals.empty(); }
     size_t size() const { return vals.size(); }
-    std::unordered_map<std::string, std::any> vals;
+    std::unordered_map<std::string, BlackboardValue> vals;
+    YAML::Node node;
+
+private:
+    friend BlackboardUtil;
+    YAML::Node getBlackboardValueNode(std::string key) const
+    {
+        for (const auto& entry : node) {
+            if (entry[Strings::key].as<std::string>() == key) {
+                return entry;
+            }
+        }
+        return YAML::Node();
+    }
+
+    std::string getBlackboardValueType(const std::string& key) const
+    {
+        YAML::Node entry = getBlackboardValueNode(key);
+        return entry[Strings::stateType].as<std::string>();
+    }
+
+    template <class T>
+    std::string getNameFromType() const;
+};
+
+struct Converter
+{
+    static constexpr const char* typeNames[] = {"bool", "int64", "double", "std::string", "std::any"};
+
+    // YAML doesnt support std::any, needs a separate list of types for iteration
+    static constexpr const char* typeNamesYAML[] = {"bool", "int64", "double", "std::string"};
+    using TypesYAML = std::variant<bool, int64_t, double, std::string>;
+
+    // convert from type to name as string
+    template <class T>
+    static constexpr const char* typeName()
+    {
+        return typeNames[BlackboardValue{T{}}.index()];
+    }
+
+    // set the value given the type name as a string
+    static void setDefaultValue(const std::string& key, const std::string& typeName, const YAML::Node& defaultValue, BlackboardImpl& bb);
+
+    // used to set the value without a yaml node
+    static void setValue(const std::string& key, const BlackboardValue& value, const std::string& typeName, BlackboardImpl& bb);
+
+    template <class T>
+    struct YamlAs
+    {
+        static T as(const YAML::Node& value) { return value.as<T>(); }
+    };
+
+    template <class T, std::size_t INDEX, template <class> class Parser = YamlAs>
+    struct makeVariantIfEqual
+    {
+        static BlackboardValue make(std::size_t index, const YAML::Node& value)
+        {
+            if (index == INDEX) {
```

Comment:
this is not required right? Can just do
`return index == INDEX ? BlackboardValue{Parser<T>::as(value)} : BlackboardValue{std::any{}};`